library/core/src/main/java/com/google/android/exoplayer2/offline/DownloadHelper.java

```diff
@@ -723,6 +727,61 @@ public DownloadRequest getDownloadRequest(String id, @Nullable byte[] data) {
     return new DownloadRequest(id, downloadType, uri, streamKeys, cacheKey, data);
   }
 
+  /**
+   * Try to estimate the total size of chunked downloads. Returns {@link C#LENGTH_UNSET} for
+   * progressive downloads.
+   * @param estimatedAudioBitrateWhenUnknown Used when audio track bitrate is not known, typical with HLS
+   * alternate audio tracks.
+   * @param estimatedTextBytesPerMinute Estimated average bytes per minute in text tracks.
+   * @return Estimated total download size, in bytes, or {@link C#LENGTH_UNSET} if not known.
+   */
+  public long estimateTotalSize(int estimatedAudioBitrateWhenUnknown, int estimatedTextBytesPerMinute) {
+
+    if (DownloadRequest.TYPE_PROGRESSIVE.equals(downloadType)) {
+      return C.LENGTH_UNSET;
+    }
+
+    assertPreparedWithMedia();
+
+    @Nullable final Object manifest = getManifest();
+    final long durationMs;
+    if (manifest instanceof DurationProvider) {
+      durationMs = ((DurationProvider) manifest).getDurationMs();
+      if (durationMs == C.TIME_UNSET) {
+        return C.LENGTH_UNSET;
+      }
+    } else {
+      return C.LENGTH_UNSET;
+    }
+
+    long selectedSize = 0;
+
+    for (int periodIndex = 0; periodIndex < this.getPeriodCount(); periodIndex++) {
+
+      final MappingTrackSelector.MappedTrackInfo mappedTrackInfo = mappedTrackInfos[periodIndex];
+      final int rendererCount = mappedTrackInfo.getRendererCount();
+
+      for (int rendererIndex = 0; rendererIndex < rendererCount; rendererIndex++) {
+        final List<TrackSelection> trackSelections = this.getTrackSelections(periodIndex, rendererIndex);
+        for (TrackSelection selection : trackSelections) {
+          final Format format = selection.getSelectedFormat();
+
+          int bitrate = format.bitrate;
+          if (bitrate == Format.NO_VALUE) {
```

Comment:
As noted in the generic comment, please leave out unspecified bit rates as it likely only applies to the following cases:
(a) tracks that have very little data compared to audio and video (e.g. text and metadata tracks)
(b) tracks that are actually muxed into other tracks and are already accounted for (e.g. muxed HLS audio tracks). 

Or to put it another way: If at least one of the `Format`s specifies a bitrate, it's very likely that all relevant non-negligible bitrates are specified in other `Format`s as well. After all, it's only an estimate :) 

If you want to add a fixed offset for the text size, you could also do that in your app afterwards using the app-specific knowledge and assumptions you have about the size of the text data. If you need to know the duration of the stream, you can probably add a new method `DownloadHelper.getTimeline()` similiar to the existing `getManifest` method to allow access to duration information.

Reply:
I agree, but since `DownloadHelper` does not provide the application with a duration, it can't even do the estimation -- even if it does know that the audio bitrate in HLS is 128k, for example.

Maybe an alternative is to add a `getDuration()` method so the app can do the estimation, but even then - the app has to know (or "remember") how many audio tracks were selected for download. Remember this is mainly for HLS with alternate audio streams, where bitrate is not specified (it's not even in the spec).

The trouble with text tracks, BTW, is that they are small (especially if they they are gzip-compressed), and the http request overhead (without http/2) is the most significant factor in their download speed. This is particularly important when the text tracks themselves are chunked. This estimation does not "fix" this problem, but it might allow some mitigation because the app developer can "play" with the given value until it makes sense.

Reply:
@tonihei your thoughts on adding `DownloadHelper.getDurationUs()` (or `Ms`), please. Note that it will have to iterate through all periods, so maybe it should be cached in a field.

Reply:
An application can probably get the duration fairly easily by calling `getManifest()` and casting to the type of media that it's working with.

Reply:
I suppose it can, but having to cast is not ideal. And still, it would have to count audio tracks that have no bitrate across all periods. Like I wrote, this is an HLS issue.

Reply:
> your thoughts on adding DownloadHelper.getDurationUs() (or Ms), please.

Could you just deprecate getManifest and add getTimeline instead? The Timeline contains both the manifest and the duration information. 

> The trouble with text tracks, BTW, is that they are small (especially if they they are gzip-compressed), and the http request overhead (without http/2) is the most significant factor in their download speed. 

That's indeed a problem when it comes to download speed, but it doesn't influence the download size as it is calculated here.

> how many audio tracks were selected for download

I see the issue now. Originally I was assuming that you just download the muxed audio stream and that its bitrate is already part of the muxed stream bitrate. But adding alternative audio streams indeed complicates the matter. 

Could you add a small inner class `interface FormatBitrateEstimator{ int estimateFormatBitrate(Format format); }` that is passed into this method? And then also add a version of the method without parameters which uses the default implementation of `format -> format.bitrate == NO_VALUE ? 0 : format.bitrate`. This ensures that the basic method returns a sensible estimate and for more advanced use cases where you care about the formats without bitrates (as you do), you can use the extended method. Having the interface keeps it flexible to also include metadata tracks etc. Would that work for you?

Reply:
Implemented all suggestions.